### ISSUE TITLE
XMP and IPTC image metadata support

### ISSUE THREAD

**Issue by LNA-DEV** [gohugoio/hugo#13146]
Hej. First of all thanks for this amazing project!

As said in #12651 the new metadata package supports XMP and IPTC metadata. Would be really nice to have this feature therefore I created this issue.

Is there any update / plan on this topic?

**Comment by bep** [comment]
>Is there any update / plan on this topic?

Not really. The main challenge with this, as I remember back when I implemented the new library, was how to represent all these tags in a data structure. This gets incredible confusing when you start looking into it; grouping the tags by their source (EXIF, XMP, IPTC) may be fine in Adobe Lightroom, but I'm not sure it's very user friendly from Hugo ... But as I said, I have not thought about this lately.

**Comment by LNA-DEV** [comment]
I fully get your point. It really gets complicated if you think about it. 

I thought maybe we could do some sort of guessing logic (for the most-used properties) **and** include a raw section so the user can, if they want to, fall back to it. But I’m not sure if a guessing logic of any sort is a good idea. You could do something like checking EXIF first and if null, checking IPTC and if that’s also null, taking the XMP property. But this could get a bit confusing if there are different values set in those sources. If well-documented and communicated, though, this might be a reasonable compromise. 

But maybe you or someone else has a better idea^^

Anyway, I’m looking forward to it :smile:

**Comment by LNA-DEV** [comment]
I have experimented with this a bit today. I think it would be most useful if we do something like I did in this pull request: https://github.com/gohugoio/hugo/pull/13539

This way existing pages still work the same but the user can find additional tags in the Exif object under Tags. (I am not sure here if IPTC and XMP should be place together with EXIF under a own Metadata object. But I think this could lead to confusion and could break support for existing sites. So I did not do it that way.)

Under Tags are all tags from every metadata category with the priority Exif > IPTC > XMP but if someone wants a specific tag from a specific category this is still possible because of the "raw" lists of tags.

**Comment by solonovamax** [comment]
> This way existing pages still work the same but the user can find additional tags in the Exif object under Tags. (I am not sure here if IPTC and XMP should be place together with EXIF under a own Metadata object. But I think this could lead to confusion and could break support for existing sites. So I did not do it that way.)

honestly, I think something like the following would work (and doesn't have any backwards compat issues):
```go
// ExifInfo holds the decoded Exif data for an Image.
type ExifInfo struct {
	// GPS latitude in degrees.
	Lat float64
	// GPS longitude in degrees.
	Long float64
	// Image creation date/time.
	Date time.Time

	// A collection of common tags for this Image.
	Tags Tags

	// A collection of the raw EXIF, IPTC, and XMP tags for this Image.
	Raw RawTags
}

type RawTags struct {
	// A collection of the available EXIF tags for this Image.
	Exif Tags

	// A collection of the available IPTC tags for this Image.
	Iptc Tags

	// A collection of the available XMP tags for this Image.
	Xmp Tags
}
```

the reason I suggest having a `RawTags` struct, is that accessing the a tag by doing something like `$image.Exif.Raw.Exif.LensModel`

however, imo the best way to do it would be to leave `ExifInfo` as-is and instead deprecate for removal in the future, and introduce a new field `Metadata ImageMetadata` onto the images, which looks something like the following:

```go
type ImageMetadata struct {
	// Image title.
	Title string
	// Image orientation. ("horizontal", "mirrored horizontal", "rotate 180", "mirrored vertical", "mirrored horizontal, rotate 270 CW", "rotate 90 CW", "mirror horizontal, rotate 90 CW", "rotate 270 CW")
	Orientation string
	// Image author.
	Author string
	// Image description.
	// note: not sure if "Description" or "AltText" is a better name here.
	Description string
	// Image copyright notice.
	CopyrightNotice string
	// Image creation date/time.
	DateCreated time.Time
	// Image modification date/time.
	DateModified time.Time
	// The person who edited this Image.
	ImageEditor string
	// The software used to edit this Image.
	ImageEditingSoftware string

	// The raw Exif, IPTC, and XMP tags for this Image.
	Raw RawTags
}

type CameraMetadata struct {
	CameraMake string
	CameraModel string
	Exposure time.Duration
	FNumber float64
	FocalLength float64
	ISO int64
	ShutterSpeed time.Duration
	Aperture float64
	Brightness float64
	Lens string
	LensMake string
	LensModel string
	Photographer string
	Gamma float64
}

type GPSMetadata struct {
	// GPS latitude in degrees.
	Lat float64
	// GPS longitude in degrees.
	Long float64
	// GPS altitude in meters.
	Altitude float64
	// GPS speed in m/s.
	Speed float64
}

type RawTags struct {
	// A collection of the raw EXIF tags for this Image.
	Exif Tags

	// A collection of the raw IPTC tags for this Image.
	Iptc Tags

	// A collection of the raw XMP tags for this Image.
	Xmp Tags
}
```

so, for example, you'd get the alt text for an image using `$image.Metadata.Description`.

with the properties sourced from the following (for this, I'm using the [exiv2 names](https://exiv2.org/metadata.html)), in order of priority:
- `ImageMetadata`:
  - `Title`: `Exif.Image.DocumentName`, `Exif.Photo.ImageTitle`, `Xmp.dc.title`, `Xmp.iptcExt.ArtworkOrObject.AOTitle`, `Exif.Image.ImageDescription` (yes, the EXIF spec says to treat `ImageDescription` like a title?)
  - `Orientation`: `Exif.Image.Orientation`, `Iptc.Application2.ImageOrientation`, `Xmp.tiff.Orientation`
  - `Author`: `Exif.Image.Artist`, `Xmp.dc.creator`, `Iptc.Application2.Byline`, `Exif.Photo.ImageEditor`, `Xmp.iptcExt.ArtworkOrObject.AOCurrentCopyrightOwnerName`, `Xmp.iptcExt.ArtworkOrObject.AOCreator`, `Xmp.plus.ImageCreatorName`, `Xmp.acdsee.author`
  - `Description`: `Xmp.iptc.ExtDescrAccessibility`, `Xmp.iptc.AltTextAccessibility`, `Xmp.dc.description`, `Iptc.Application2.Caption`, `Exif.Photo.UserComment`, `Exif.Image.ImageDescription`, `Iptc.Application2.Headline`, `Xmp.iptcExt.ArtworkOrObject.AOContentDescription`, `Xmp.acdsee.caption`
  - `CopyrightNotice`: `Exif.Image.Copyright`, `Xmp.dc.rights`, `Iptc.Application2.Copyright`, `Xmp.iptcExt.ArtworkOrObject.AOCopyrightNotice`
  - `DateCreated`: `Exif.Image.DateTimeOriginal` + `Exif.Photo.SubSecTimeOriginal`, `Xmp.photoshop.DateCreated`, `Xmp.xmp.CreateDate`, `Iptc.Application2.DateCreated` + `Iptc.Application2.TimeCreated`, `Xmp.iptcExt.ArtworkOrObject.AODateCreated`, Xmp.
  - `DateModified`: `Exif.Image.DateTime` + `Exif.Photo.SubSecTime` (yes `DateTime` is actually the modification date & time), `Xmp.xmp.ModifyDate`
  - `ImageEditor`: `Exif.Photo.ImageEditor`
  - `ImageEditingSoftware`: `Exif.Photo.ImageEditingSoftware`, `Exif.Image.Software`, `Exif.Image.ProcessingSoftware`, `Xmp.xmp.CreatorTool`
- `CameraMetadata`:
  - `CameraMake`: `Exif.Image.Make`, `Xmp.tiff.Make`
  - `CameraModel`: `Exif.Image.Model`, `Exif.Image.UniqueCameraModel`, `Xmp.tiff.Model`
  - `Exposure`: `Exif.Image.ExposureTime`, `Xmp.crs.Exposure`
  - `FNumber`: `Exif.Image.FNumber`
  - `ISO`: `Exif.Image.ISOSpeedRatings`, `Exif.Photo.ISOSpeed` (tbh not sure the difference between this and the previous one. the first comes from before Exif 2.3, and the second was added in Exif 2.3), `Xmp.exif.ISOSpeedRatings`, `Xmp.exifEX.ISOSpeed`
  - `ShutterSpeed`: `Exif.Image.ShutterSpeedValue`, `Xmp.exif.ShutterSpeedValue`
  - `Aperture`: `Exif.Image.ApertureValue`, `Xmp.exif.ApertureValue`
  - `Brightness`: `Exif.Image.BrightnessValue`, `Xmp.exif.BrightnessValue`, `Xmp.crs.Brightness`
  - `FocalLength`: `Exif.Image.FocalLength`, `Xmp.exif.FocalLength`
  - `Lens`: `Exif.Photo.LensSpecification`, `Xmp.aux.Lens`
  - `LensMake`: `Exif.Photo.LensMake`, `Xmp.exifEX.LensMake`
  - `LensModel`: `Exif.Photo.LensModel`, `Xmp.exifEX.LensModel`
  - `Photographer`: `Exif.Photo.Photographer`
  - `Gamma`: `Exif.Photo.Gamma`
- `GPSMetadata`
  - `Lat`: `Exif.GPSInfo.GPSLatitude`, `Xmp.exif.GPSLatitude`
  - `Long`: `Exif.GPSInfo.GPSLongitude`, `Xmp.exif.GPSLongitude`
  - `Altitude`: `Exif.GPSInfo.GPSAltitude`, `Xmp.exif.GPSAltitude`
  - `Speed`: `Exif.GPSInfo.GPSSpeed`, `Xmp.exif.GPSSpeed`

however, all of the different XMP tags sound like a pain to work with, so I could understand excluding them.

and, with the field being named `Metadata`, in the future you might be able to introduce maybe a `VideoMetadata` type or an `AudioMetadata` type, which all use the name `Metadata` and it just depends on what type the resource is.

**Comment by bep** [comment]
OK, I'm going to tackle this now, with a compromise that reduces the complexity of the implementation 
/maintenance/documentation:

* Leave `Image.Exif` as is, possibly deprecate in the future.
* Add `Image.Meta` that returns the struct below.
* Adjust the `imaging` config to enable one or more of these tag sources (XMP comes with a performance penalty, which may be important to some).
* Note that there's no `AllTags` concept in that struct, but it's hard to know upfront how to merge these -- so people can easly do that with `{{ $tags := merge .Exif .IPTC .XMP }}`

Sounds ... good enough?

```go
// MetaInfo holds the decoded metadata for an Image.
// Note that the diffeerent sources needs to be handled separately in the config.
// If you want a consolidated view of the different tag sections, use the merge template func, e.g. {{ $m := merge .Exif .IPTC .XMP }}.
// TODO1 add a test for merge.
type MetaInfo struct {
	// GPS latitude in degrees.
	// TODO1 probably need to look in all sections for for lat, long and date, until we find a value
	Lat float64

	// GPS longitude in degrees.
	Long float64

	// Image creation date/time.
	Date time.Time

	Exif Tags
	IPTC Tags
	XMP  Tags
}

```

/cc @LNA-DEV and others.

**Comment by solonovamax** [comment]
> * Note that there's no `AllTags` concept in that struct, but it's hard to know upfront how to merge these -- so people can easly do that with `{{ $tags := merge .Exif .IPTC .XMP }}`

yeah, merging them is a rather tedious process (see: the many different tags I went through for each of the properties I listed)

> Sounds ... good enough?

I think it would be nice to also possibly include the title, author, description/alt, and copyright notice. possible just the Exif & IPTC values, as those are easiest to parse and there's not too many of them.
though I also get not wanting to maintain all of that
